Northstar.CustomServers/mod/scripts/vscripts/gamemodes/_gamemode_at.nut

```diff
@@ -16,302 +58,1563 @@ const array<string> VALID_BOUNTY_TITAN_SETTINGS = [
+			string campEntVarName = "camp" + string( spawnId + 1 ) + "Ent"
+			bool waveNotActive = GetGlobalNetBool( "preBankPhase" ) || GetGlobalNetBool( "banksOpen" )
+			if ( !IsValid( GetGlobalNetEnt( campEntVarName ) ) && !waveNotActive )
+				SetGlobalNetEnt( campEntVarName, CreateCampTracker( curCampData, spawnId ) )
+			
+			array<AT_SpawnData> minionSquadDatas
+			foreach ( AT_SpawnData data in curSpawnData )
+			{
+				switch ( data.aitype )
+				{
+					case "npc_soldier":
+					case "npc_spectre":
+					case "npc_stalker":
+						if ( !AT_USE_TOTAL_ALLOWED_ON_FIELD_CHECK )
+							minionSquadDatas.append( data )
+						else
+							thread AT_DroppodSquadEvent_Single( curCampData, spawnId, data )
+						break
+					
+					case "npc_super_spectre":
+						thread AT_ReaperEvent( curCampData, spawnId, data )
+						break
+				}
+			}
+
+			// minions squad spawn
+			if ( !AT_USE_TOTAL_ALLOWED_ON_FIELD_CHECK )
+			{
+				if ( minionSquadDatas.len() > 0 )
+					thread AT_DroppodSquadEvent( curCampData, spawnId, minionSquadDatas )
+			}
+
+			// use campProgressThink for handling wave state
+			thread CampProgressThink( spawnId, totalNPCsToSpawn )
+		}
+		else // bosswave spawn
+		{
+			foreach ( AT_SpawnData data in curSpawnData )
+			{
+				if( data.aitype != "npc_titan" )
+					continue
+				
+				thread AT_BountyTitanEvent( curCampData, spawnId, data )
+				break
+			}
+		}
+	}
+}
+
+void function CampProgressThink( int spawnId, int totalNPCsToSpawn )
+{
+	string campLetter = GetCampLetter( spawnId )
+	string campProgressName = campLetter + "campProgress"
+	string campEntVarName = "camp" + string( spawnId + 1 ) + "Ent"
+
+	// initial wait
+	SetGlobalNetFloat( campProgressName, 1.0 )
+
+	// TODO: random wait, make this a constant ??
+	wait 3.0
+
+	float cleanUpTime = -1.0
+
+	while ( true )
+	{
+		int npcsLeft
+		// get all npcs might be in this camp
+		for ( int i = 0; i < 5; i++ )
+		{
+			string netVarName = string( i + 1 ) + campLetter + "campCount"
+			int netVarValue = GetGlobalNetInt( netVarName )
+			if ( netVarValue >= 0 ) // uninitialized network var starts from -1, avoid checking them
+				npcsLeft += netVarValue
+		}
+		//print( "npcsLeft: " + string( npcsLeft ) )
+
+		float campLeft = float( npcsLeft ) / float( totalNPCsToSpawn )
+		//print( "campLeft: " + string( campLeft ) )
+		SetGlobalNetFloat( campProgressName, campLeft )
+
+		if( npcsLeft <= AT_CAMP_BORED_NPCS_LEFT_TO_START_CLEANUP && cleanUpTime < 0.0 )
+		{
+			cleanUpTime = Time() + AT_CAMP_BORED_CLEANUP_WAIT
+			print("Cleanup timer started!")
+		}
+
+		if( Time() > cleanUpTime && cleanUpTime > 0.0 && spawnId in file.campScriptEntArrays )
+		{
+			foreach( int handle in file.campScriptEntArrays[spawnId] )
+			{
+				array<entity> entities = GetScriptManagedEntArray( handle )
+				entities.removebyvalue( null )
+				foreach ( entity ent in entities )
+				{
+					if ( IsAlive( ent ) && ent.IsNPC() )
+						ent.Die()
```

Comment:
```suggestion
					if ( IsAlive( ent ) && ent.IsNPC() )
					{
						printt( "Killing bored AI " + ent.GetClassName() + " at " + ent.GetOrigin() )
						ent.Die()
					}
```
I would like to log some details about bored/stuck AI, should help a bit with improving navmeshes in the future